### DISCUSSION TITLE
Contributing a Rust API for mosaik

### DISCUSSION THREAD

**Discussion by lu391see** [OFFIS-mosaik/mosaik#15] [Q&A] [answered]
Hi everyone,
we are currently developing a mosaik API for Rust. For now it's a private GitLab repo but we are happy to make it available on GitHub.

We want to get in touch with you to talk about a potential release process into your project and further questions like:
- Should we create an Issue on GitLab or communicate via the GitHub Discussion section?
- What are the preferred ways of contributing the API? i.e. handing over the code via a PR or linking our repository in the documentation and keep it in a self-owned repo.
- Which functionalities are the minimum requirements? The async functions are not supported from our side yet, as well as `finalize()` and `configure()`.
- Some questions about the API typing. Especially where the types.py contradicts the official documentation, e.g. this [given examples](https://mosaik.readthedocs.io/en/latest/mosaik-api/high-level.html#mosaik_api_v3.Simulator.get_data).
- Who to talk to for further questions?
- In the [docs](https://mosaik.gitlab.io/api/MosaikAPIv3.jl/#async-requests) of the newly released Julia API it is stated that the async functions `set_data()` and `get_data()` are deprecated in mosaik3. Where do we find further info like that?

Happy to hear from you.
Kind regards,
Lukas

**Comment by eike** [comment]
Hi Lukas,

that is great to hear! APIs for additional programming languages are always welcome (and it might give me a reason to finally have a more serious look at Rust). To your points:

- For now, the discussions here are the way to go. More specific discussions on technical details should probably be done in the adapter's repository at some point.
- Both options are fine for us, and it mostly depends on how much "control" you want to keep over the API. If you want us to manage it, the requirements would be that you are happy with the MIT license. We would then create a repository in the mosaik group on GitLab, and you could create Merge Requests into it. Otherwise (if you prefer a different license or want the repository to be your own), we are also happy with just linking to it in our docs.
- In general, the functionality of the low-level mosaik API should be supported, in my opinion. This would include `finalize`, but not `configure`. Configuring the simulator should be done in a way that is most natural for the host programming language (i.e. Rust). (As you have looked at the Julia API, notice that we also deviated from the Python procedure there.)
- Yeah, the types are less perfect than I would like. mosaik was created pre-types in Python, so sometimes giving absolutely correct types is infeasible. (I assume that you are talking about the `"time"` field in `get_data`'s output, for which I have not found a good solution yet.) In general, my recommendation would be to try and make the API idiomatic as possible for the host language. As Rust has much stronger typing discipline, I would try to design the Rust API with sensible types, and translate to the slightly messy JSON as the very last step behind the scenes. (In my experience, this can go endlessly deep, though. Be as pragmatic as necessary.)
- I think I would be the right person to talk to. You can just keep commenting here, and I will reply, you can send emails to mosaik@offis.de (or directly to me, eike.schulte@offis.de), and if you want, we can also set up a video call at some point.
- You are right, this is not well-documented. (If you use these methods in a scenario, you will get a deprecation warning.) We should add that to the docs.

Best regards
Eike

**Reply by lu391see** [reply]
Hi Eike,
thank you for your response. I have two follow-up questions:

- I don't see a `finalize` function in the [low-level API documentation](https://mosaik.readthedocs.io/en/latest/mosaik-api/low-level.html), but I can find it in the [high-level API documentation](https://mosaik.readthedocs.io/en/latest/mosaik-api/high-level.html). Is this what you meant or do you mean the `setup_done`? Could you please explain how/if they are connected, as they are distinct functions in the high-level API and what functionality would be expected of the Rust API?
- Should we support the async functions, before contributing the API to the mosaik project? This is not yet implemented by us.

Kind regards
Lukas

**Reply by eike** [reply]
- You are right. I don't know the tool myself, apparently. :grimacing: In other APIs, `finalize` is called after the simulator receives the low-level `stop` message from mosaik and no communication with mosaik is possible at that point, anymore. Simulators do occasionally have a need for a little bit of tear-down at the end (flushing buffers, etc.), but I will retract my requirement that this happen in a `finalize` method.
- I think support for `get_related_entities`, `get_progress` and `set_event` would be good. Feel free to skip the asynchronous `set_data` and `get_data` calls (but not the normal `get_data` mosaik method, of course), as we do in the Julia API. That said, I'm also fine with a pre-1.0 version missing those features.

**Reply by lu391see** [reply, marked as answer]
Our mosaik API for Rust was released to [crates.io](https://crates.io/crates/mosaik-api) (The Rust community's package directory) and the code is open sourced on [Github](https://github.com/ekut-es/mosaik-api). Feel free to have a look :)

**Reply by eike** [reply]
This is great to hear, thanks. Unless you object, we would add links to this in our documentation.

(I have only had a cursory glance, so I cannot say with absolute confidence that everything is correct from mosaik's perspective, but what I saw looked reasonable. I should really take this as an excuse to write some simulator in Rust myself.)

**Reply by lu391see** [reply]
Of course, go ahead! It would make me very happy if someone finds this useful and builds on it!
